dace/runtime/include/dace/cuda/dynmap.cuh

```diff
@@ -22,228 +22,205 @@
+                        index_type tmp = warp.shfl_up(prefix, delta);
+                        prefix += (warpThreadRank >= delta) ? tmp : 0;
                     }
-                    if (np_shared.warp.owner[warp_id] == lane_id)
-                    {
-                        // This thread is owner 
-                        np_shared.warp.start[warp_id] = local_start;
-                        np_shared.warp.size[warp_id] = local_size;
-
-                        // Mark this work-item as processed   
-                        local_start = 0;
-                        local_size = 0;
+
+                    // total number and local start index (need to subtract localSize, because it is including the
+                    // local element)
+                    index_type total = warp.shfl(prefix, warpSize - 1);
+                    index_type warp_offset = warpId * WARP_SIZE*WARP_SIZE;
+                    index_type thread_offset = prefix - localSize;
+                    index_type offset = warp_offset + thread_offset;
+
+                    // write element range and owner to shared memory
+                    for (int i = 0; i < localSize; i++) {
+                        s.fg.src[offset + i] = localSrc;
+                        s.fg.data[offset + i] = localStart + i;
+                    }
+
+                    warp.sync(); // necessary because warps may diverge more since volta
+
+                    // do work
+                    for (unsigned int j = warp_offset + warpThreadRank; j < warp_offset + total; j += warpSize) {
+                        work(s.fg.src[j], s.fg.data[j]);
+                    }
+
+                } else {
+
+                    warp.sync();
+
+                    // do work
+                    if (localSize > 0) {
+                        for (int j = localStart; j < localEnd; j++) {
+                            work(localSrc, j);
+                        }
                     }
-                    start = np_shared.warp.start[warp_id];
-                    size = np_shared.warp.size[warp_id];
-                    */
-                }
 
-                for (int ii = lane_id; ii < size; ii += WP_SIZE)
-                {
-                    work(start + ii, src);
                 }
-            }
 
-            __syncthreads();
+            } else {
 
-            //
-            // Third scheduler: tackle all work-items with size < 32 serially   
-            //
-            // It is possible to disable this scheduler by setting NP_WP_CROSSOVER to 0 
+                if (localSize > 0) {
+                    // do work
+                    for (int j = localStart; j < localEnd; j++) {
+                        work(localSrc, j);
+                    }
+
+                }
 
-            for (int ii = 0; ii < local_size; ii++)
-            {
-                work(local_start + ii, local_src);
             }
+
         }
+
     };
 
-}  // namespace dace
+}
+
```

Comment:
quite a bit of whitespace here between the closing braces. I would also consider putting comments next to the "else"s to say which scheduler we are ending/ignoring